File: 프로그래머스/lv2/49993. 스킬트리/스킬트리.cpp
#include <string>
#include <vector>
#include <unordered_map>

using namespace std;

bool checkSkillTree(unordered_map<char, int>& m, string& st) {
    int idx = 0;
    for (auto c : st) {
        auto it = m.find(c);
        if (it != m.end()) {
            if (it->second == idx) {
                idx++;
                if (idx == m.size() - 1) return true;
            }
            else return false;
        }
    }
    return true;
}

int solution(string skill, vector<string> skill_trees) {
    unordered_map<char, int> m;
    for (int i = 0; i < skill.size(); i++) m.insert({skill[i], i});
    int cnt = 0;
    for (auto st : skill_trees) {
        if (checkSkillTree(m, st)) cnt++;
    }
    return cnt;
}

File: 프로그래머스/2/49993. 스킬트리/스킬트리.java
import java.util.*;

class Solution {
    
    private Set<Character> skillSet = new HashSet<>();
    
    private boolean checkSkillTree(String skill, String st) {
        int idx = 0;
        for (int i = 0; i < st.length(); i++) {
            char c = st.charAt(i);
            if (skillSet.contains(c)) {
                if (idx >= skill.length() || skill.charAt(idx) != c) 
                    return false;
                else idx++;
            }
        }
        return true;
    }
    
    public int solution(String skill, String[] skill_trees) {
        for (int i = 0; i < skill.length(); i++)
            skillSet.add(skill.charAt(i));
        int cnt = 0;
        for (String skillTree : skill_trees)
            if (checkSkillTree(skill, skillTree)) cnt++;
        return cnt;
    }
}